pkg/server/server_controller_sql.go

```diff
@@ -117,35 +116,28 @@ func (c *serverController) shouldWaitForTenantServer(name roachpb.TenantName) bo
 func (c *serverController) waitForTenantServer(
 	ctx context.Context, name roachpb.TenantName,
 ) (onDemandServer, error) {
-	// Note that requests that come in after the first request may
-	// time out in less time than the
-	// WaitForClusterStartTimeout. This seems fine for now since
-	// cluster startup should be relatively quick and if it isn't
-	// often waiting longer isn't going to help.
+	// Note that requests that come in after the first request may time out
+	// in less time than the WaitForClusterStartTimeout. This seems fine for
+	// now since cluster startup should be relatively quick and if it isn't,
+	// waiting longer isn't going to help.
 	s, _, err := c.tenantWaiter.Do(ctx, string(name), func(ctx context.Context) (interface{}, error) {
-		// TODO(ssd): Ideally, we could avoid this polling by having
-		// the tenant controller hand us a channel that it will close
-		// when the tenant becomes available.
-		maxWait := multitenant.WaitForClusterStartTimeout.Get(&c.st.SV)
-		retryCfg := retry.Options{
-			InitialBackoff: 50 * time.Millisecond,
-			MaxBackoff:     1 * time.Second,
-		}
-
-		ctx, cancel := context.WithTimeout(ctx, maxWait)
-		defer cancel()
-
-		var (
-			err error
-			s   onDemandServer
-		)
-		for i := retry.StartWithCtx(ctx, retryCfg); i.Next(); {
-			s, err = c.getServer(ctx, name)
+		t := timeutil.NewTimer()
+		defer t.Stop()
+		t.Reset(multitenant.WaitForClusterStartTimeout.Get(&c.st.SV))
+		for {
+			s, waitCh, err := c.getServer(ctx, name)
 			if err == nil {
-				break
+				return s, nil
+			}
```

Comment:
we should only wait if the err matches `server for tenant %q not ready`, correct? If so,  I feel like this error string should have its own mark.

Reply:
No, we want to retry on both errors. Both the existence of the tenant and the startup is async from the perspective of the server controller because we learn about tenants via a rangefeed.